### PR TITLE
fix(wallet): remove the generic from wallet

### PR DESCRIPTION
**Pull request by rustaceanrob**
### Description

The `PersistenceBackend` uses generics to describe errors returned while applying the change set to the persistence layer. This change removes generics wherever possible and introduces a new public error enum. Removing the generics from `PersistenceBackend` errors is the first step towards #1363 

*Update*: I proceeded with removing the generics from `Wallet` by introducing a `Box<dyn PersistenceBackend>` . 

### Notes to the reviewers

This one sort of blew up in the number of changes due to the use of generics for most of the `Wallet` error variants. The generics were only used for the persistence errors, so I removed the generics from higher level errors whenever possible. The error variants of `PersistenceBackend` may also be more expressive, but I will level that up for discussion and make any changes required.

### Changelog notice

- Changed `PersistenceBackend` errors to depend on the `anyhow` crate.
- Remove the generic `T` from `Wallet`

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

### CODE REVIEW

**Comment by evanlinjin** [comment]
Thanks for working on this. In the future, please state in the ticket that you'll be working on it, otherwise we'll result in duplicate work.

**Comment by rustaceanrob** [comment]
Will do! Thanks for the tip

**Comment by rustaceanrob** [comment]
It is not obvious why this change is useful for removing `D` from Wallet. I am going to continue work towards that goal on this branch.

File: crates/chain/src/persist.rs
Comment (notmandatory) [inline crates/chain/src/persist.rs]: 🔧 This description doesn't capture that this can also be thrown for load errors. How about something simple like:
```suggestion
/// Errors returned by [`PersistBackend`] when writing or loading data.
```

Comment (rustaceanrob) [reply crates/chain/src/persist.rs]: Agreed. I am also working on more expressive error variants by wrapping the errors.

**Comment by rustaceanrob** [comment]
Broke no_std with an error variant. Will fix soon.

File: crates/chain/src/persist.rs
Comment (LLFourn) [inline crates/chain/src/persist.rs]: Shouldn't we include the `std::io::Error` here? The variant will need to be `#[cfg(feature = "std")]` of course.

File: crates/chain/src/persist.rs
Comment (LLFourn) [inline crates/chain/src/persist.rs]: This error variant doesn't make sense to me since it only applies to backends that iter (also it's not clear what the caller is meant to do with this error).

File: crates/chain/src/persist.rs
Comment (LLFourn) [inline crates/chain/src/persist.rs]: Should probably have different error enum for loading/appending.

File: crates/chain/src/persist.rs
Comment (rustaceanrob) [reply crates/chain/src/persist.rs]: Yes we should. I was unaware you are able to feature gate variants like this. The marco just goes above the variant?

**Comment by LLFourn** [comment]
So after looking and thinking about this more closely I'd expect two error enums, one for each interaction:

```rust
enum LoadError {
   #[cfg(feature = "std")]
   StdIo(std::io::Error),
   Other(String),
}

enum CommitError {
    #[cfg(feature = "std")]
   StdIo(std::io::Error),
   Other(String),
   Inconsistent(String) // maybe?
}
```

But looking at it we don't really get much value from this structure unless we add specific error variants that occur with our changesets that we can report and react to. But our changeset API is (mostly) makes invalid representations impossible. In so far as invalid representations are possible they depend on the concrete changeset which the backend doesn't have access to.

The correct thing to do here I think is to return `Box<dyn core::error::Error>` (except that doesn't exist yet!). How about we just use `anyhow::Error` to do this. It would be better to not have this dependency but I think replicating what `anyhow` does (by `Box`ing our own error trait) is the next best solution and seems like a waste of time when almost every rust app depends on anyhow anyway. I think we should just accept it and wait for `core::error::Error`.

cc @evanlinjin

File: crates/chain/src/persist.rs
Comment (rustaceanrob) [reply crates/chain/src/persist.rs]: This error was intended to capture the error in `file_store/store.rs`. As far as I understand, not all users of `bdk_wallet` will use `file_store`, so I wasn't sure how to wrap it in the `PersistBackendError`. In `store`, this error occurs when applying the change set, so I guess `ApplyChangeError` could specify there is something wrong when attempting that action?

**Comment by rustaceanrob** [comment]
Added my interpretation of how we would use `anyhow` here. Easy enough to add on or revert back.

File: crates/file_store/src/store.rs
Comment (evanlinjin) [inline crates/file_store/src/store.rs]: Instead of changing the signature of this method, could we just map the error when we use it in `PersistBackend::write_changes`?

File: crates/file_store/src/store.rs
Comment (evanlinjin) [inline crates/file_store/src/store.rs]: Nit: could we either revert this change, or use `fmt::Formatter`/`fmt::Result`/`fmt::Display`?

File: crates/chain/src/persist.rs
Comment (LLFourn) [inline crates/chain/src/persist.rs]: `stdwith` D:

**Review by LLFourn — COMMENTED** [review]
This looks great. Main question is whether we're comfortable with this in `bdk_chain` or we should just make a new `bdk_persist` crate which defines this trait. It is rather orthogonal to everything else in `bdk_chain` and now it adds another dependency. If `bdk_core` existed (#1155) likely it wouldn't even depend on `bdk_chain`. "persist" could also be a feature flag in `bdk_chain`.

As a side note, for some reason we use this trait in the `bdk_chain` examples but I'm not sure why. They could easily be rewritten to not depend on this trait.

cc @thunderbiscuit wdyt about `anyhow::Error`. How do you handle errors in the FFI?

File: crates/bdk/src/wallet/mod.rs
Comment (evanlinjin) [inline crates/bdk/src/wallet/mod.rs:368]: We might need to update the docs as this variant can represent either a writing or loading error now.

File: crates/bdk/src/wallet/mod.rs
Comment (evanlinjin) [inline crates/bdk/src/wallet/mod.rs]:
```suggestion
            NewOrLoadError::Persist(e) => write!(f, "failed to either write to or load from persistence, {}", e),
```

File: crates/bdk/src/wallet/mod.rs
Comment (evanlinjin) [inline crates/bdk/src/wallet/mod.rs]: Should we hide the `Box` from the caller and call `Box::new` internally? 
```suggestion
        db: impl PersistBackend<ChangeSet> + 'static,
```

File: crates/chain/src/persist.rs
Comment (evanlinjin) [inline crates/chain/src/persist.rs]: I think it is nicer to hide the `Box` from the caller here. I can't think of a benefit for exposing it?

```suggestion
    pub fn new(backend: impl PersistBackend<C> + 'static) -> Self {
```

**Review by evanlinjin — CHANGES_REQUESTED** [review]
This looks great! However, I'm wondering whether it's better to introduce our own lightweight trait instead on using `anyhow` (which is another dependency).

```rust
pub trait PersistError: core::any::Any + Debug {}

pub fn an_example_method() -> Result<(), Box<dyn PersistError>> { todo!() }
```

The downsides of this is that the `PersistenceBackend` implementations may need to wrap their returned errors in a new type to implement `PersistError`. However, I see this as a minor compromise as they need to implement `PersistenceBackend` anyway?

cc. @LLFourn @ValuedMammal

File: crates/bdk/src/wallet/mod.rs
Comment (evanlinjin) [reply crates/bdk/src/wallet/mod.rs:368]: @LLFourn what are your thoughts about combining write/load variants?

**Comment by rustaceanrob** [comment]
Committed those changes for review

File: crates/bdk/src/wallet/mod.rs
Comment (LLFourn) [reply crates/bdk/src/wallet/mod.rs:368]: Fine to keep them as one variant but as you said doc strings need to be updated if so. I don't see a strong motivation to merge them.

File: crates/bdk/Cargo.toml
Comment (LLFourn) [inline crates/bdk/Cargo.toml]: version should just be ` = "1" `

**Comment by LLFourn** [comment]
> The downsides of this is that the `PersistenceBackend` implementations may need to wrap their returned errors in a new type to implement `PersistError`. However, I see this as a minor compromise as they need to implement `PersistenceBackend` anyway?
> 
> cc. @LLFourn @ValuedMammal

I don't think this is a minor downside for the implementer. It means wrapping all errors from the underlying backend in a type and implementing a trait on the wrapped types. Also then what about downcasting? What about backtraces? `anyhow::Error` just does everything right and there's no point in trying to replicate it. 

I'm leaning towards putting the persist trait in its own crate `bdk_persist`. It never made much sense in `bdk_chain`, it was just convenient to put it there but it no longer is.

File: crates/bdk/src/wallet/mod.rs
Comment (rustaceanrob) [reply crates/bdk/src/wallet/mod.rs:368]: For the comment above the variant, how is "Either writing to or loading from the persistence backend failed." Verbatim of the `NewOrLoadError::Persist` error message.

**Comment by rustaceanrob** [comment]
Should I continue by breaking out into a bdk persist crate? Or is that best left for another PR?

**Comment by ValuedMammal** [comment]
> The correct thing to do here I think is to return `Box<dyn core::error::Error>` (except that doesn't exist yet!). How about we just use `anyhow::Error` to do this. It would be better to not have this dependency but I think replicating what `anyhow` does (by `Box`ing our own error trait) is the next best solution and seems like a waste of time when almost every rust app depends on anyhow anyway. I think we should just accept it and wait for `core::error::Error`.
> 

I'd be in favor of finding a solution that doesn't require anyhow, if possible. Currently, the `PersistBackend` errors are only constrained by `Debug`, so a new `PersistError` struct (or enum) could be something like

```rust
pub struct PersistError(pub Box<dyn fmt::Debug + Send + Sync>);
```

and make `PersistBackend` implementations return this concrete error, which still feels a bit backwards, but in practice could be as simple as writing a `From` conversion or constructing the type directly.

The persist module could also just live in `bdk` crate along with wallet, no?

**Comment by evanlinjin** [comment]
@rustaceanrob we haven't reached a consensus on this. Feel free to put in your input.

**Comment by LLFourn** [comment]
> > The correct thing to do here I think is to return `Box<dyn core::error::Error>` (except that doesn't exist yet!). How about we just use `anyhow::Error` to do this. It would be better to not have this dependency but I think replicating what `anyhow` does (by `Box`ing our own error trait) is the next best solution and seems like a waste of time when almost every rust app depends on anyhow anyway. I think we should just accept it and wait for `core::error::Error`.
> 

Replicating what anyhow does 100% is definitely not in scope. You can attach context anyhow errors and backtraces are automatically attached if your rust version is new enough. Of course we don't need this but it is nice to have.

> I'd be in favor of finding a solution that doesn't require anyhow, if possible. Currently, the `PersistBackend` errors are only constrained by `Debug`, so a new `PersistError` struct (or enum) could be something like
> 
> ```rust
> pub struct PersistError(pub Box<dyn fmt::Debug + Send + Sync>);
> ```
> 

And also `Display`. But we need to go further and make it `+ core::any::Any + 'static` so you can downcast the error to the concrete one. You will likely want to get the specific backend error at the application level e.g. the concrete postgres error. Then impl `From<E> where: E: std::error::Error`. This is quite a lot of stuff to maintain to avoid a dependency which virtually every application uses (and should use!). The question boils down to whether we want to maintain our own dynamic error system when there is already a standard choice for this very common problem in the rust ecosystem. I'd say no but it's less my decision to make than those who will actually be doing the work :)

> and make `PersistBackend` implementations return this concrete error, which still feels a bit backwards, but in practice could be as simple as writing a `From` conversion or constructing the type directly.
> 
> The persist module could also just live in `bdk` crate along with wallet, no?

Yes that could also work but I have a feeling the persist trait will be updated much less frequently than `bdk_wallet`. It wouldn't make sense to have to bump the version on the sqlite crate because some orthogonal API change happened in `bdk_wallet`.

**Comment by thunderbiscuit** [comment]
> cc @thunderbiscuit wdyt about anyhow::Error. How do you handle errors in the FFI?

I'm not familiar with anyhow but reading up on it now. In general I favour not reimplementing the wheel, and think that using tried-and-true libraries is more robust over time (i.e., I am in favour of using anyhow instead of building similar behaviour in-house). One question I'd ask is if we add the anyhow dependency here, does it mean we'll have 2 approaches to errors across the codebase? Or would it be used in this single instance only?

As for FFI, we can't pass the errors directly (very!) unfortunately, so we roughly have to create new ones and map the Rust errors to those. This change would not impact us too much either way; do whatever is best for the Rust side of things and we'll adapt.

**Comment by rustaceanrob** [comment]
I understand the reluctance to bring in a new dependency, but as @LLFourn described, the behavior that we are looking for would just be a re-implementation of what `anyhow` already does. `anyhow` also adds an additional end-user benefit of context and backtraces that feel out of scope for this PR, but improve the end-user developer experience imo. I am personally not seeing enough drawbacks compared to both the end-user and maintainer benefits gained from a drop in replacement, but I am new here of course! 

On structure, I am not seeing `chain` import `bdk`, so I'm not sure about moving it to the wallet side of things. This is type of common behavior that should rarely change seems like a prime candidate for `bdk_core`. If I were to weigh in on structure, I would leave things how they are for this PR and wait for a definitive restructuring decision, or start off `bdk_core` with `bdk_core/persist`. 

> As for FFI, we can't pass the errors directly (very!) unfortunately, so we roughly have to create new ones and map the Rust > errors to those. This change would not impact us too much either way; do whatever is best for the Rust side of things and > we'll adapt.

I have contributed to code with public `anyhow` result types along with UniFFI bindings. I will reach out to him to see what his experience was. Happy to help with UniFFI chores.

**Review by LLFourn — APPROVED** [review]
Fine to leave it here for now.  I still have a weak preference for moving it into `bdk_persist` rather than `bdk_core` (don't really want `bdk_core` to depend on `anyhow`). 

ACK bfadae799bad187b42732614b5c8738c5eb26819

Might be nice to clean up the commit history before merging. I think this should just be one commit on top of master if we can manage it.

File: example-crates/example_cli/src/lib.rs
Comment (ValuedMammal) [inline example-crates/example_cli/src/lib.rs]: nit: No need for this type alias now, just use `Persist` everywhere.

File: crates/bdk/src/wallet/mod.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs]: Instead of importing this globally, can we just prefix the `Result`s that use it with `anyhow::`. That way it's easy to spot where it is being used.

File: crates/bdk/Cargo.toml
Comment (ValuedMammal) [inline crates/bdk/Cargo.toml:1]: anyhow is in here twice, so I guess we can drop it from dev-dependencies?

File: crates/chain/src/persist.rs
Comment (ValuedMammal) [inline crates/chain/src/persist.rs:16]: Alternatively, we can define the trait as

```rust
pub trait PersistBackend<C>: fmt::Debug {...}
```

and still be able to `#[derive(Debug)]` as before. I think you'd have to add a bound `C: fmt::Debug` for example in the trait impl for `Store`. Feel free to ignore if you think it's inconsequential.

**Review by ValuedMammal — COMMENTED** [review]
I acknowledge all the points in the previous discussion, and I can offer these review comments :)

File: crates/chain/src/persist.rs
Comment (rustaceanrob) [reply crates/chain/src/persist.rs:16]: Just going to leave as is to keep my change minimal for this commit

File: crates/bdk/Cargo.toml
Comment (rustaceanrob) [reply crates/bdk/Cargo.toml:1]: Dropped. Will adjust if CI fails. The library uses `anyhow = { version = "1", default-features = false }`, so hopefully the CI doesn't use any additional features.

**Comment by rustaceanrob** [comment]
Should I squash commits locally? I only ask because I think there's also a feature to do it on the merge GUI.

File: crates/bdk/Cargo.toml
Comment (ValuedMammal) [reply crates/bdk/Cargo.toml:1]: yep I see now. maybe better to keep it as dev dependency as well in order to get default features.

File: crates/bdk/src/keys/mod.rs
Comment (ValuedMammal) [inline crates/bdk/src/keys/mod.rs]: nit: Seems unnecessary

File: crates/bdk/src/wallet/error.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/error.rs]: Do we actually need this? I was able to build and test without it.

File: crates/bdk/src/wallet/mod.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs]: nit: `DescriptorError` looks nicer.

File: crates/file_store/src/store.rs
Comment (ValuedMammal) [inline crates/file_store/src/store.rs]: I would use `anyhow` macro.

```suggestion
        self.append_changeset(changeset)
            .map_err(|e| anyhow!(e).context("failed to write changes to persistence backend"))
```

File: crates/file_store/src/store.rs
Comment (ValuedMammal) [inline crates/file_store/src/store.rs]: We still want to map to `iter_error`.

```suggestion
        self.aggregate_changesets()
            .map_err(|e| anyhow!(e.iter_error).context("error loading from persistence backend"))
```

**Review by ValuedMammal — CHANGES_REQUESTED** [review]
Sorry, I have a couple more notes.

Primary suggestions are for `store.rs`, and I'm also wondering whether the `From` impl in `wallet::error` module is needed. Any nits are definitely not blockers.

As far as squashing, my preference is to use [`git rebase --interactive`](https://git-scm.com/docs/git-rebase).

File: crates/bdk/src/wallet/error.rs
Comment (rustaceanrob) [reply crates/bdk/src/wallet/error.rs]: When previously adding new variants I thought it might be useful. Agree that this doesn't make much sense to add at the moment.

**Comment by rustaceanrob** [comment]
Squashed to [c958fb4](https://github.com/bitcoindevkit/bdk/commit/c958fb468702203baf7e1c8f5f0ac8d89d3b9505) with changes suggested from @ValuedMammal. Should be ready for CI

File: crates/chain/src/persist.rs
Comment (jirijakes) [inline crates/chain/src/persist.rs]: The documentation is outdated now, `B` no more.

**Review by jirijakes — COMMENTED** [review]
This change silently removes `Sync` and `Send` auto traits from `Persist` and therefore from anything that uses it (e. g. `Wallet`). 

One way to deal with it could be:

```rs
pub struct Persist<C> {
    backend: Box<dyn PersistBackend<C> + Send + Sync>,
    stage: C,
}
```

but that would require every backend to be `Send` + `Sync`, even when not needed.

I am not sure how big of a problem this is, but given that `Wallet` until now tried to be `Sync` + `Send`, it may be an issue.

**Comment by rustaceanrob** [comment]
As far as I can tell `Wallet` is already [not `Sync`](https://docs.rs/bdk/latest/bdk/wallet/struct.Wallet.html#synthetic-implementations)? Let me know if I am misunderstanding something!

**Comment by jirijakes** [comment]
> As far as I can tell `Wallet` is already not `Sync`?

It is in [alpha.8](https://docs.rs/bdk/1.0.0-alpha.8/bdk/wallet/struct.Wallet.html#impl-Sync-for-Wallet%3CD%3E).

But I really don't know whether this was some accident or deliberate action. I don't know whether this is something to care about or not.

**Comment by rustaceanrob** [comment]
Interesting. I tried messing with it and I am not aware of any other work-arounds besides the one you proposed. Between `Sync` and `Send`, not having `Send` seems to be the most problematic as earlier releases of `Wallet` were `Send`. Is it reasonable for all backends to be `Send` and `Sync`? @LLFourn @evanlinjin

**Comment by ValuedMammal** [comment]
That's a great point @jirijakes what made you realize that? I mean is it evident from looking at the code, or is there something more subtle going on? I do think it's generally desirable to maintain thread safety of the wallet.

**Comment by jirijakes** [comment]
> what made you realize that?

Rather a coincidence. A few days ago I read articles (mainly [this](https://predr.ag/blog/semver-in-rust-tooling-breakage-and-edge-cases/)) about this kind of semver breakage. Having the articles fresh in mind and seeing `Box`, I was curious.

**Comment by rustaceanrob** [comment]
I updated the PR to include the `Sync + Send` trait bounds on the `PersistBackend` when creating a new `Wallet`. Although this may be a breaking change for wallets that implement their own store, they should easily be able to enforce `Sync + Send` on their `ChangeSet` generic. The benefit gained of `Wallet` being thread safe feels like a reasonable tradeoff, especially in the server context. Thank you @jirijakes for that insight.

**Comment by LLFourn** [comment]
Yeah this change  implies that we must force `PersistBackend` to be `Send + Sync`. Before `Send` and `Sync` depended on `PersistBackend` impl. Now since we are using a trait object we have to make a decision for the user and the right decision is to add `Send + Sync` to the trait object. It's a slight regression but easily worth it for what we get. For things that are not `Send` or `Sync` you can always just make a channel or something like that to communicate with it (where the channel sender would be `Send + Sync`).

**Review by evanlinjin — APPROVED** [review]
ACK e51af49ffa951e14203ac2b465ade351dd90f6cd

This is good work. However, I'm really not a fan of introducing the `anyhow` dependency to the `bdk_chain` crate.

I think we should create a separate RP which introduces a separate `bdk_persist` crate.

**Pull request merged**
